File: backend/flask/routes/suscriptor.py
from flask import Blueprint, request, jsonify
import hashlib
from functions.db_admin import (
    readCanciones,
    readAlbumes,
    readArtistas,
    getIdArtistaCancion,
)
from controller.db_user import (
    getPerfil,
    passwordCorrecto,
    modificarPerfil,
    buscarCanciones,
    buscarAlbumes,
    buscarArtistas,
    favorito,
    getFavoritos,
    getIdPlaylist,
    createPlaylist,
    readPlaylists,
    addCancionPlaylist,
    deleteCancionPlaylist,
    readCancionesPlaylist,
    getTopCanciones,
    getTopArtistas,
    getTopAlbums,
    getHistorial,
    reproducirAlbum,
    reproducirArtista,
    reproducirAleatorio,
)
from mis3.config_s3 import guardarImagen

suscriptor_routes = Blueprint("suscriptor_routes", __name__)

@suscriptor_routes.route("/inicio", methods=["GET"])
def inicio():
    try:
        c = readCanciones()
        alb = readAlbumes()
        art = readArtistas()

        return jsonify({"canciones": c["canciones"], "albums": alb["albums"], "artistas": art["artistas"]})
    except Exception as error:
        print(error)
        return jsonify({"canciones": [], "albums": [], "artistas": []})

@suscriptor_routes.route("/perfil", methods=["POST"])
def perfil():
    try:
        id_usuario = request.json.get("id_usuario")
        perfil = getPerfil(id_usuario)
        return jsonify(perfil)
    except Exception as error:
        print(error)
        return jsonify({"imagen": "", "nombre": "", "apellido": "", "email": ""})

@suscriptor_routes.route("/modificar-perfil", methods=["POST"])
def modificar_perfil():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        imagen = data["imagen"]
        nombre = data["nombre"]
        apellido = data["apellido"]
        email = data["email"]
        password = data["password"]
        correcto = passwordCorrecto(id_usuario, hashlib.sha256(password.encode()).hexdigest())
        
        if correcto["status"]:
            result = modificarPerfil(id_usuario, nombre, apellido, email)
            if imagen != "":
                guardarImagen("usuarios/" + str(id_usuario), imagen)
                result["ok"] = True
            return jsonify(result)
        return jsonify({"ok": False}), 400
    except Exception as error:
        print(error)
        return jsonify({"ok": False}), 400

@suscriptor_routes.route("/buscar", methods=["POST"])
def buscar():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        buscar = data["buscar"]
        c = buscarCanciones(id_usuario, buscar)
        alb = buscarAlbumes(buscar)
        art = buscarArtistas(buscar)

        return jsonify({"canciones": c["canciones"], "albums": alb["albums"], "artistas": art["artistas"]})
    except Exception as error:
        print(error)
        return jsonify({"ok": False}), 400

@suscriptor_routes.route("/favorito", methods=["POST"])
def marcar_favorito():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        fav = data["fav"]
        result = favorito(id_usuario, fav)
        return jsonify(result)
    except Exception as error:
        print(error)
        return jsonify({"ok": False}), 400

@suscriptor_routes.route("/favorites", methods=["POST"])
def obtener_favoritos():
    try:
        id_usuario = request.json.get("id_usuario")
        result = getFavoritos(id_usuario)
        return jsonify(result)
    except Exception as error:
        print(error)
        return jsonify({"songs": []})

@suscriptor_routes.route("/playlist", methods=["POST"])
def crear_playlist():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        nombre = data["nombre"]
        descripcion = data["descripcion"]
        imagen = data["imagen"]
        existente = getIdPlaylist(id_usuario, nombre)
        
        if not existente["status"]:
            result = createPlaylist(id_usuario, nombre, descripcion, imagen)
            if result["status"]:
                return jsonify(result["listado_playlists"])
        return jsonify({"ok": False, "playlist": []})
    except Exception as error:
        print(error)
        return jsonify({"ok": False, "playlist": []})

@suscriptor_routes.route("/playlists", methods=["POST"])
def obtener_playlists():
    try:
        id_usuario = request.json.get("id_usuario")
        result = readPlaylists(id_usuario)
        return jsonify(result)
    except Exception as error:
        print(error)
        return jsonify({"playlist": []})

@suscriptor_routes.route("/playlist/add-song", methods=["POST"])
def agregar_cancion_a_playlist():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        id_cancion = data["id_cancion"]
        nombre_playlist = data["nombre_playlist"]
        playlist = getIdPlaylist(id_usuario, nombre_playlist)
        
        if playlist["status"]:
            result = addCancionPlaylist(id_cancion, playlist["id_playlist"])
            return jsonify(result)
        return jsonify({"songs": []})
    except Exception as error:
        print(error)
        return jsonify({"songs": []})

@suscriptor_routes.route("/playlist/delete-song", methods=["POST"])
def eliminar_cancion_de_playlist():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        id_cancion = data["id_cancion"]
        nombre_playlist = data["nombre_playlist"]
        playlist = getIdPlaylist(id_usuario, nombre_playlist)
        
        if playlist["status"]:
            result = deleteCancionPlaylist(id_cancion, playlist["id_playlist"])
            return jsonify(result)
        return jsonify({"songs": []})
    except Exception as error:
        print(error)
        return jsonify({"songs": []})

@suscriptor_routes.route("/inplaylist", methods=["POST"])
def obtener_canciones_de_playlist():
    try:
        data = request.json
        id_usuario = data["id_usuario"]
        nombre = data["nombre"]
        playlist = getIdPlaylist(id_usuario, nombre)
        
        if playlist["status"]:
            result = readCancionesPlaylist(playlist["id_playlist"])
            urlImagen = f"{prefijoBucket}Fotos/playlists/{playlist['id_playlist']}.jpg"
            return jsonify({"songs": result["canciones"], "imagen_playlist": urlImagen})
        else:
            print(f'Playlist "{nombre}" de usuario con ID {id_usuario} no encontrada.')
        return jsonify({"songs": [], "imagen_playlist": ""})
    except Exception as error:
        print(error)
        return jsonify({"songs": [], "imagen_playlist": ""})

@suscriptor_routes.route("/historial", methods=["POST"])
def obtener_historial():
    try:
        id_usuario = request.json.get("id_usuario")
        topC = getTopCanciones(id_usuario)
        topA = getTopArtistas(id_usuario)
        topAlb = getTopAlbums(id_usuario)
        historialC = getHistorial(id_usuario)
        
        return jsonify({
            "cancionesRep": topC["canciones"],
            "artistaRep": topA["artistas"],
            "albumRep": topAlb["albums"],
            "historial": historialC["canciones"]
        })
    except Exception as error:
        print(error)
        return jsonify({
            "cancionesRep": [],
            "artistaRep": [],
            "albumRep": [],
            "historial": []
        })

@suscriptor_routes.route("/reproducir", methods=["POST"])
def reproducir():
    data = request.json
    id_usuario = data["id_usuario"]
    id = data["id"]
    tipo = data["tipo"]
    try:
        if tipo == 0:
            resultArtista = getIdArtistaCancion(id)
            if resultArtista["status"]:
                result = reproducirArtista(id_usuario, resultArtista["id_artista"])
                return jsonify(result)
        elif tipo == 1:
            result = reproducirAlbum(id_usuario, id)
            return jsonify(result)
        elif tipo == 2:
            result = reproducirArtista(id_usuario, id)
            return jsonify(result)
        return jsonify({"tracks": []})
    except Exception as error:
        print(error)
        return jsonify({"tracks": []})

@suscriptor_routes.route("/reproducir-aleatorio", methods=["POST"])
def reproducir_aleatorio():
    try:
        id_usuario = request.json.get("id_usuario")
        result = reproducirAleatorio(id_usuario)
        return jsonify(result)
    except Exception as error:
        print(error)
        return jsonify({"tracks": []})

File: backend/node/routes/suscriptor.js
var router = require('express').Router();
const sha256 = require('js-sha256');
const { readCanciones, readAlbumes, readArtistas, getIdArtistaCancion } = require('../controller/db_admin')
const { getPerfil, passwordCorrecto, modificarPerfil, buscarCanciones, buscarAlbumes, buscarArtistas,
        favorito, getFavoritos,
        getIdPlaylist, createPlaylist, readPlaylists, addCancionPlaylist, deleteCancionPlaylist, readCancionesPlaylist,
        getTopCanciones, getTopArtistas, getTopAlbums, getHistorial,
        reproducirAlbum, reproducirArtista, reproducirAleatorio } = require('../controller/db_user');
const { guardarImagen } = require('../controller/s3');
const prefijoBucket = process.env.PREFIJO_BUCKET;

router.get('/inicio', async (req, res) => {
    try {
        const c = await readCanciones();
        const alb = await readAlbumes();
        const art = await readArtistas();
        
        res.status(200).json({ canciones: c.canciones, albums: alb.albums, artistas: art.artistas });
    } catch (error) {
        console.log(error);
        res.status(400).json({canciones:[]}, {albums:[]}, {artistas:[]});
    }
});

router.post('/perfil', async (req, res) => {
    try {
        const id_usuario = req.body.id_usuario;
        const perfil = await getPerfil(id_usuario);
        res.status(200).json(perfil)
    } catch (error) {
        console.log(error);
        res.status(400).json({ imagen: '', nombre: '', apellido: '', email: '' })
    }
});

router.post('/modificar-perfil', async (req, res) => {
    try {
        const { id_usuario, imagen, nombre, apellido, email, password } = req.body;
        const correcto = await passwordCorrecto(id_usuario, sha256(password));
        if (correcto.status) {
            const result = await modificarPerfil(id_usuario, nombre, apellido, email);
            if (imagen != "") {
                guardarImagen('usuarios/' + id_usuario, imagen);
                result.ok = true;
            }
            return res.status(200).json(result);
        }
        res.status(400).json({ ok : false });
    } catch (error) {
        console.log(error);
        res.status(400).json({ ok : false });
    }
});

router.post('/buscar', async (req, res) => {
    try {
        const { id_usuario, buscar } = req.body;
        const c = await buscarCanciones(id_usuario, buscar);
        const alb = await buscarAlbumes(buscar);
        const art = await buscarArtistas(buscar);

        res.status(200).json({ canciones: c.canciones, albums: alb.albums, artistas: art.artistas });
    } catch (error) {
        console.log(error);
        res.status(400).json({ ok: false });
    }
});

router.post('/favorito', async (req, res) => {
    try {
        const { id_usuario, fav } = req.body;
        const result = await favorito(id_usuario, fav);
        res.status(result.ok? 200: 400).json(result);
    } catch (error) {
        console.log(error);
        res.status(400).json({ ok: false });
    }
});

router.post('/favorites', async (req, res) => {
    try {
        const id_usuario = req.body.id_usuario;
        const result = await getFavoritos(id_usuario);
        res.status(200).json(result);
    } catch (error) {
        console.log(error);
        res.status(400).json({ songs: [] });
    }
});

router.post('/playlist', async (req, res) => {
    try {
        const { id_usuario, nombre, descripcion, imagen } = req.body;
        
        const existente = await getIdPlaylist(id_usuario, nombre);
        if (!existente.status) {
            const result = await createPlaylist(id_usuario, nombre, descripcion, imagen);
            if (result.status) {
                return res.status(200).json(result.listado_playlists);
            }
        }
        res.status(400).json({ok: false, playlist: [] });
    } catch (error) {
        console.log(error);
        res.status(400).json({ ok: false, playlist: [] })
    }
});

router.post('/playlists', async (req, res) => {
    try {
        const id_usuario = req.body.id_usuario;
        const result = await readPlaylists(id_usuario);
        res.status(200).json(result);
    } catch (error) {
        console.log(error);
        res.status(400).json({ playlist: [] })
    }
});

router.post('/playlist/add-song', async(req, res) => {
    try {
        const { id_usuario, id_cancion, nombre_playlist } = req.body;
        const playlist = await getIdPlaylist(id_usuario, nombre_playlist);
        if (playlist.status) {
            const result = await addCancionPlaylist(id_cancion, playlist.id_playlist);
            return res.status(200).json(result);
        }
        res.status(400).json({songs: []});
    } catch (error) {
        console.log(error);
        res.status(400).json({ songs: [] })
    }
});

router.post('/playlist/delete-song', async(req, res) => {
    try {
        const { id_usuario, id_cancion, nombre_playlist } = req.body;
        const playlist = await getIdPlaylist(id_usuario, nombre_playlist);
        if (playlist.status) {
            const result = await deleteCancionPlaylist(id_cancion, playlist.id_playlist);
            return res.status(200).json(result);
        }
        res.status(400).json({songs: []});
    } catch (error) {
        console.log(error);
        res.status(400).json({ songs: [] })
    }
});

router.post('/inplaylist', async(req, res) => {
    try {
        const { id_usuario, nombre } = req.body;
        const playlist = await getIdPlaylist(id_usuario, nombre);
        if (playlist.status) {
            const result = await readCancionesPlaylist(playlist.id_playlist);
            const urlImagen = `${prefijoBucket}Fotos/playlists/${playlist.id_playlist}.jpg`;
            return res.status(200).json({ songs: result.canciones, imagen_playlist: urlImagen });
        } else { 
            console.log(`Playlist "${nombre}" de usuario con ID ${id_usuario} no encontrada.`);
        }
        res.status(400).json({ songs: [], imagen_playlist: '' });
    } catch (error) {
        console.log(error);
        res.status(400).json({ songs: [], imagen_playlist: '' })
    }
});

router.post('/historial', async(req, res) => {
    try {
        const id_usuario = req.body.id_usuario;
        const topC = await getTopCanciones(id_usuario);
        const topA = await getTopArtistas(id_usuario);
        const topAlb = await getTopAlbums(id_usuario);
        const historialC = await getHistorial(id_usuario);
        return res.status(200).json({ cancionesRep: topC.canciones, artistaRep: topA.artistas,
                                      albumRep: topAlb.albums, historial: historialC.canciones });
    } catch (error) {
        console.log(error);
        res.status(400).json({ cancionesRep: [], artistaRep: [], albumRep: [], historial: [] })
    }
});

router.post('/reproducir', async (req, res) => {
    const { id, tipo, id_usuario } = req.body;
    try {
        if (tipo == 0) {//Canciones
            const resultArtista = await getIdArtistaCancion(id);
            if (resultArtista.status) {
                const result = await reproducirArtista(id_usuario, resultArtista.id_artista);
                return res.status(200).json(result);
            }
        } else if (tipo == 1) {//Album
            const result = await reproducirAlbum(id_usuario, id);
            return res.status(200).json(result);
        } else if (tipo == 2) {//Artista
            const result = await reproducirArtista(id_usuario, id);
            return res.status(200).json(result);
        }
        res.status(400).json({ tracks: [] });
    } catch {
        console.log(error);
        res.status(400).json({ tracks: [] });
    }

});

router.post('/reproducir-aleatorio', async (req, res) => {
    try {
        const id_usuario = req.body.id_usuario;
        const result = await reproducirAleatorio(id_usuario);
        res.status(200).json(result);
    } catch (error) {
        console.log(error);
        res.status(400).json({ tracks: [] });
    }
});

module.exports = router;